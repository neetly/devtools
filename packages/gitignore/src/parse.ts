export const parse = (content: string): { patterns: readonly string[] } => {
  const patterns = content
    .split(/\r?\n/)

    // Remove comments
    .filter((pattern) => !pattern.startsWith("#"))
    .map((pattern) => pattern.replace(/^\\#/, "#"))

    // Remove trailing spaces and empty lines
    .map((pattern) => pattern.replace(/(?<!\\) +$/, "").replace(/\\ $/, " "))
    .filter((pattern) => pattern !== "")

    // Remove non-sense patterns
    .filter((pattern) => !/^!?\/?$/.test(pattern))

    // Normalize slashes
    .map((pattern) => {
      const isNegative = pattern.startsWith("!");
      let path = pattern.replace(/^!/, "");
      if (!/^\/|\/(?!$)/.test(path)) {
        path = "**/" + path;
      }
      path = path.replace(/^\//, "");
      return (isNegative ? "!" : "") + path;
    })

    // Escape special characters
    .map((pattern) => pattern.replace(/[(){}\\]/g, "\\$&"));

  return { patterns };
};
