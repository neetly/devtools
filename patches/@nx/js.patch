diff --git a/src/plugins/typescript/plugin.js b/src/plugins/typescript/plugin.js
index edbf6c01d8c0d42b4c396bd11954cc99382f67f2..9a4678cbf2b9f00ec0859c5d4689724cb05a0083 100644
--- a/src/plugins/typescript/plugin.js
+++ b/src/plugins/typescript/plugin.js
@@ -205,7 +205,6 @@ function getInputs(namedInputs, configFilePath, tsConfig, internalProjectReferen
     extendedConfigFiles.files.forEach((configPath) => {
         configFiles.add(configPath);
     });
-    externalDependencies.push(...extendedConfigFiles.packages);
     const includePaths = new Set();
     const excludePaths = new Set();
     const projectTsConfigFiles = [
