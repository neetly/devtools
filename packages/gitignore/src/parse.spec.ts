import { describe, expect, test } from "@jest/globals";
import { Minimatch } from "minimatch";

import { parse } from "./parse.js";

const createTester = (content: string) => {
  const { patterns } = parse(content);
  const minimatches = patterns.map((pattern) => {
    return new Minimatch(pattern, { dot: true });
  });

  const matches = (path: string) => {
    let isMatched = false;
    for (const minimatch of minimatches) {
      if (minimatch.negate) {
        isMatched &&= minimatch.match(path);
      } else {
        isMatched ||= minimatch.match(path);
      }
    }
    return isMatched;
  };

  return {
    patterns,

    ignores: (path: string) => {
      const components = path.split("/");
      for (let index = 0; index < components.length; index++) {
        const directory = components.slice(0, index).join("/") + "/";
        if (matches(directory)) {
          return true;
        }
      }
      return matches(path);
    },
  };
};

describe("parse", () => {
  test.each([
    // BOM
    ["\uFEFFfile", { file: true }],

    // LF and CRLF
    ["file\ndir/file", { file: true, "dir/file": true }],
    ["file\r\ndir/file", { file: true, "dir/file": true }],

    // comments
    ["#file", { "#file": false }],
    ["\\#file", { "#file": true }],

    // spaces
    ["  ", { " ": false, "  ": false }],
    [" \\ ", { " ": false, "  ": true }],
    ["\\  ", { " ": true, "  ": false }],
    ["  file", { file: false, " file": false, "  file": true }],
    ["fi  le", { file: false, "fi le": false, "fi  le": true }],
    ["file  ", { file: true, "file ": false, "file  ": false }],
    ["file\\  ", { file: false, "file ": true, "file  ": false }],
    ["file \\ ", { file: false, "file ": false, "file  ": true }],

    // empty and non-sense patterns
    ["", { file: false }],
    ["!", { file: false }],
    ["/", { file: false }],
    ["!/", { file: false }],

    // negative patterns
    ["!file", { file: false, "!file": false }],
    ["\\!file", { file: false, "!file": true }],

    // slashes ("/") and two consecutive asterisks ("**")
    ["file", { file: true, "sub/file": true }],
    ["/file", { file: true, "sub/file": false }],
    ["**/file", { file: true, "sub/file": true }],

    ["dir/file", { "dir/file": true, "sub/dir/file": false }],
    ["/dir/file", { "dir/file": true, "sub/dir/file": false }],
    ["**/dir/file", { "dir/file": true, "sub/dir/file": true }],

    ["dir", { "dir/file": true, "sub/dir/file": true }],
    ["/dir", { "dir/file": true, "sub/dir/file": false }],
    ["**/dir", { "dir/file": true, "sub/dir/file": true }],

    ["dir/", { "dir/file": true, "sub/dir/file": true }],
    ["/dir/", { "dir/file": true, "sub/dir/file": false }],
    ["**/dir/", { "dir/file": true, "sub/dir/file": true }],

    ["**", { file: true, "sub/file": true, "sub/sub/file": true }],
    ["/**", { file: true, "sub/file": true, "sub/sub/file": true }],
    ["**/", { file: false, "sub/file": true, "sub/sub/file": true }],
    ["/**/", { file: false, "sub/file": true, "sub/sub/file": true }],

    ["/dir/**", { "dir/file": true, "dir/sub/file": true }],
    ["/dir/**/", { "dir/file": false, "dir/sub/file": true }],
    ["/dir/**/file", { "dir/file": true, "dir/sub/file": true }],

    [["/dir/**", "!/dir/file"], { "dir/file": false }],
    [["/dir/**", "!/dir/sub/file"], { "dir/sub/file": true }],
    [["/dir/**", "!/dir/sub", "!/dir/sub/file"], { "dir/sub/file": false }],
    [["/dir/**", "!/dir/sub/", "!/dir/sub/file"], { "dir/sub/file": false }],

    [["/dir/**/", "!/dir/sub"], { "dir/sub/file": false }],
    [["/dir/**/", "!/dir/sub/"], { "dir/sub/file": false }],
    [["/dir/**/", "!/dir/sub/file"], { "dir/sub/file": true }],

    // globs
    ["a*", { a: true, ab: true, abc: true }],
    ["*c", { c: true, bc: true, abc: true }],
    ["a*c", { ac: true, abc: true, abbc: true, "a/c": false }],
    [
      "a/*/c",
      { "a/c": false, "a/b/c": true, "a/bb/c": true, "a/b/b/c": false },
    ],
    ["[abc]", { a: true, b: true, c: true, "[abc]": false }],
    ["[a-c]", { a: true, b: true, c: true, "[a-c]": false }],
    ["[[]", { "[": true, "[[]": false }],
    ["[]]", { "]": true, "[]]": false }],
    ["[[]]", { "[": false, "]": false, "[]": true, "[[]]": false }],
    ["[[\\]]", { "[": true, "]": true, "[]": false, "[[]]": false }],
    ["[{}()]", { "{": true, "}": true, "(": true, ")": true, "[{}()]": false }],
    ["{a,b,c}", { a: false, b: false, c: false, "{a,b,c}": true }],
    ["?(a|b|c)", { a: false, b: false, c: false, "?(a|b|c)": true }],
    ["*(a|b|c)", { a: false, b: false, c: false, "*(a|b|c)": true }],
    ["+(a|b|c)", { a: false, b: false, c: false, "+(a|b|c)": true }],
    ["@(a|b|c)", { a: false, b: false, c: false, "@(a|b|c)": true }],
    ["\\!(a|b|c)", { x: false, "!(a|b|c)": true }],
    ["x/!(a|b|c)", { "x/x": false, "x/!(a|b|c)": true }],

    // backslashes
    ["a\\bc", { abc: true, "a\\bc": false }],
    ["a\\\\bc", { abc: false, "a\\bc": true }],
    ["a\\*", { a: false, ab: false, abc: false, "a*": true }],
    ["\\*c", { c: false, bc: false, abc: false, "*c": true }],
    ["a\\*c", { ac: false, abc: false, abbc: false, "a*c": true }],
    ["a/\\*/c", { "a/b/c": false, "a/bb/c": false, "a/*/c": true }],
    ["\\[", { "[": true }],
    ["\\[]", { "[]": true }],
    ["\\[\\]", { "[]": true }],
    ["\\[abc", { "[abc": true }],
    ["\\[abc]", { a: false, b: false, c: false, "[abc]": true }],
    ["\\[abc\\]", { a: false, b: false, c: false, "[abc]": true }],
    ["[a\\bc]", { a: true, b: true, c: true, "\\": false }],
    ["[a\\\\bc]", { a: true, b: true, c: true, "\\": true }],
    ["[a\\-c]", { a: true, b: false, c: true, "-": true }],
  ])("pattern %j", (content, files) => {
    const { patterns, ignores } = createTester([content].flat().join("\n"));
    expect(patterns).toMatchSnapshot();
    for (const [file, isIgnored] of Object.entries(files)) {
      expect([file, ignores(file)]).toStrictEqual([file, isIgnored]);
    }
  });

  // invalid patterns
  // test.each(["[]", "[", "[abc", "[\\]", "[\\["])(
  //   "invalid pattern %j",
  //   (content) => {
  //     const { patterns } = parse(content);
  //     expect(patterns).toStrictEqual([]);
  //   },
  // );
});
