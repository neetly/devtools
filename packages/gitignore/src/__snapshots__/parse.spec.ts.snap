// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse pattern "  " 1`] = `[]`;

exports[`parse pattern "  file" 1`] = `
[
  "**/  file",
]
`;

exports[`parse pattern " \\\\ " 1`] = `
[
  "**/  ",
]
`;

exports[`parse pattern "!" 1`] = `[]`;

exports[`parse pattern "!/" 1`] = `[]`;

exports[`parse pattern "!file" 1`] = `
[
  "!**/file",
]
`;

exports[`parse pattern "" 1`] = `[]`;

exports[`parse pattern "#file" 1`] = `[]`;

exports[`parse pattern "*(a|b|c)" 1`] = `
[
  "**/*\\(a|b|c\\)",
]
`;

exports[`parse pattern "**" 1`] = `
[
  "**/*",
]
`;

exports[`parse pattern "**/" 1`] = `
[
  "**/*/",
]
`;

exports[`parse pattern "**/dir" 1`] = `
[
  "**/dir",
]
`;

exports[`parse pattern "**/dir/" 1`] = `
[
  "**/dir/",
]
`;

exports[`parse pattern "**/dir/file" 1`] = `
[
  "**/dir/file",
]
`;

exports[`parse pattern "**/file" 1`] = `
[
  "**/file",
]
`;

exports[`parse pattern "*c" 1`] = `
[
  "**/*c",
]
`;

exports[`parse pattern "+(a|b|c)" 1`] = `
[
  "**/+\\(a|b|c\\)",
]
`;

exports[`parse pattern "/" 1`] = `[]`;

exports[`parse pattern "/**" 1`] = `
[
  "**/*",
]
`;

exports[`parse pattern "/**/" 1`] = `
[
  "**/*/",
]
`;

exports[`parse pattern "/dir" 1`] = `
[
  "dir",
]
`;

exports[`parse pattern "/dir/" 1`] = `
[
  "dir/",
]
`;

exports[`parse pattern "/dir/**" 1`] = `
[
  "dir/**/*",
]
`;

exports[`parse pattern "/dir/**/" 1`] = `
[
  "dir/**/*/",
]
`;

exports[`parse pattern "/dir/**/file" 1`] = `
[
  "dir/**/file",
]
`;

exports[`parse pattern "/dir/file" 1`] = `
[
  "dir/file",
]
`;

exports[`parse pattern "/file" 1`] = `
[
  "file",
]
`;

exports[`parse pattern "?(a|b|c)" 1`] = `
[
  "**/?\\(a|b|c\\)",
]
`;

exports[`parse pattern "@(a|b|c)" 1`] = `
[
  "**/@\\(a|b|c\\)",
]
`;

exports[`parse pattern "[[\\\\]]" 1`] = `
[
  "**/[[\\]]",
]
`;

exports[`parse pattern "[[]" 1`] = `
[
  "**/[[]",
]
`;

exports[`parse pattern "[[]]" 1`] = `
[
  "**/[[]]",
]
`;

exports[`parse pattern "[]]" 1`] = `
[
  "**/[]]",
]
`;

exports[`parse pattern "[{}()]" 1`] = `
[
  "**/[\\{\\}\\(\\)]",
]
`;

exports[`parse pattern "[a\\\\\\\\bc]" 1`] = `
[
  "**/[a\\\\bc]",
]
`;

exports[`parse pattern "[a\\\\-c]" 1`] = `
[
  "**/[a\\-c]",
]
`;

exports[`parse pattern "[a\\\\bc]" 1`] = `
[
  "**/[a\\bc]",
]
`;

exports[`parse pattern "[a-c]" 1`] = `
[
  "**/[a-c]",
]
`;

exports[`parse pattern "[abc]" 1`] = `
[
  "**/[abc]",
]
`;

exports[`parse pattern "\\\\  " 1`] = `
[
  "**/ ",
]
`;

exports[`parse pattern "\\\\!(a|b|c)" 1`] = `
[
  "**/\\!\\(a|b|c\\)",
]
`;

exports[`parse pattern "\\\\!file" 1`] = `
[
  "**/\\!file",
]
`;

exports[`parse pattern "\\\\#file" 1`] = `
[
  "**/#file",
]
`;

exports[`parse pattern "\\\\*c" 1`] = `
[
  "**/\\*c",
]
`;

exports[`parse pattern "\\\\[" 1`] = `
[
  "**/\\[",
]
`;

exports[`parse pattern "\\\\[\\\\]" 1`] = `
[
  "**/\\[\\]",
]
`;

exports[`parse pattern "\\\\[]" 1`] = `
[
  "**/\\[]",
]
`;

exports[`parse pattern "\\\\[abc" 1`] = `
[
  "**/\\[abc",
]
`;

exports[`parse pattern "\\\\[abc\\\\]" 1`] = `
[
  "**/\\[abc\\]",
]
`;

exports[`parse pattern "\\\\[abc]" 1`] = `
[
  "**/\\[abc]",
]
`;

exports[`parse pattern "{a,b,c}" 1`] = `
[
  "**/\\{a,b,c\\}",
]
`;

exports[`parse pattern "a*" 1`] = `
[
  "**/a*",
]
`;

exports[`parse pattern "a*c" 1`] = `
[
  "**/a*c",
]
`;

exports[`parse pattern "a/*/c" 1`] = `
[
  "a/*/c",
]
`;

exports[`parse pattern "a/\\\\*/c" 1`] = `
[
  "a/\\*/c",
]
`;

exports[`parse pattern "a\\\\*" 1`] = `
[
  "**/a\\*",
]
`;

exports[`parse pattern "a\\\\*c" 1`] = `
[
  "**/a\\*c",
]
`;

exports[`parse pattern "a\\\\\\\\bc" 1`] = `
[
  "**/a\\\\bc",
]
`;

exports[`parse pattern "a\\\\bc" 1`] = `
[
  "**/a\\bc",
]
`;

exports[`parse pattern "dir" 1`] = `
[
  "**/dir",
]
`;

exports[`parse pattern "dir/" 1`] = `
[
  "**/dir/",
]
`;

exports[`parse pattern "dir/file" 1`] = `
[
  "dir/file",
]
`;

exports[`parse pattern "fi  le" 1`] = `
[
  "**/fi  le",
]
`;

exports[`parse pattern "file  " 1`] = `
[
  "**/file",
]
`;

exports[`parse pattern "file \\\\ " 1`] = `
[
  "**/file  ",
]
`;

exports[`parse pattern "file" 1`] = `
[
  "**/file",
]
`;

exports[`parse pattern "file\\\\  " 1`] = `
[
  "**/file ",
]
`;

exports[`parse pattern "file\\ndir/file" 1`] = `
[
  "**/file",
  "dir/file",
]
`;

exports[`parse pattern "file\\r\\ndir/file" 1`] = `
[
  "**/file",
  "dir/file",
]
`;

exports[`parse pattern "x/!(a|b|c)" 1`] = `
[
  "x/!\\(a|b|c\\)",
]
`;

exports[`parse pattern "ï»¿file" 1`] = `
[
  "**/file",
]
`;

exports[`parse pattern ["/dir/**","!/dir/file"] 1`] = `
[
  "dir/**/*",
  "!dir/file",
]
`;

exports[`parse pattern ["/dir/**","!/dir/sub","!/dir/sub/file"] 1`] = `
[
  "dir/**/*",
  "!dir/sub",
  "!dir/sub/file",
]
`;

exports[`parse pattern ["/dir/**","!/dir/sub/","!/dir/sub/file"] 1`] = `
[
  "dir/**/*",
  "!dir/sub/",
  "!dir/sub/file",
]
`;

exports[`parse pattern ["/dir/**","!/dir/sub/file"] 1`] = `
[
  "dir/**/*",
  "!dir/sub/file",
]
`;

exports[`parse pattern ["/dir/**/","!/dir/sub"] 1`] = `
[
  "dir/**/*/",
  "!dir/sub",
]
`;

exports[`parse pattern ["/dir/**/","!/dir/sub/"] 1`] = `
[
  "dir/**/*/",
  "!dir/sub/",
]
`;

exports[`parse pattern ["/dir/**/","!/dir/sub/file"] 1`] = `
[
  "dir/**/*/",
  "!dir/sub/file",
]
`;
